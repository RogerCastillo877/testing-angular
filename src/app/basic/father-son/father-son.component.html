<h4>Son</h4>
<code>{{ client | json }}</code>
<br><br>

<button *ngIf="client" class="btn btn-primary" (click)="onChange(3)" data-test="btn-change-id">Cambiar Id</button>

<button *ngIf="client" class="btn btn-danger" (click)="onDelete()" data-test="btn-delete">Delete Client</button>