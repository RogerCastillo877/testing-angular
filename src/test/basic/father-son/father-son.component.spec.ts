import { ComponentFixture, TestBed } from '@angular/core/testing';

import { FatherSonComponent } from '../../../app/basic/father-son/father-son.component';

describe('FatherSonComponent', () => {
  let component: FatherSonComponent;
  let fixture: ComponentFixture<FatherSonComponent>;
  let compiled: HTMLElement;

  beforeEach(() => {
    TestBed.configureTestingModule({
      declarations: [FatherSonComponent]
    })
      .compileComponents();
  });

  beforeEach(() => {
    fixture = TestBed.createComponent(FatherSonComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
    compiled = fixture.nativeElement;
    jest.clearAllMocks();
  })

  test('should create', () => {
    expect(component).toBeTruthy();
  });

  test('should make match with snapshot', () => {
    fixture.detectChanges();
    expect(compiled).toMatchSnapshot();
  })

  test('should not display buttons without client', () => {
    const buttons = compiled.querySelectorAll('button');
    expect(buttons.length).toBe(0)
  })

  test('should display buttons if have client', () => {
    component.client = { id: 1, name: 'Juan' };
    fixture.detectChanges();

    const buttons = compiled.querySelectorAll('button');
    expect(buttons.length).toBe(2);
  })

  test('should display 2 buttons if have client', () => {
    component.client = { id: 1, name: 'Juan' };
    fixture.detectChanges();

    expect(compiled).toMatchSnapshot;
  })

  test('should emit onDeleteClient with delete button', () => {
    component.client = { id: 1, name: 'Juan' };
    fixture.detectChanges();

    jest.spyOn(component.onDeleteClient, 'emit');

    const btnDelete = compiled.querySelector('[data-test="btn-delete"]');
    btnDelete?.dispatchEvent(new Event('click'));

    expect(component.onDeleteClient.emit).toBeCalled();
  })

  test('should emit onClientUpdated with button "change Id"', () => {
    component.client = { id: 1, name: 'Juan' };
    fixture.detectChanges();

    jest.spyOn(component.onClientUpdated, 'emit');

    const btnChangeId = compiled.querySelector('[data-test="btn-change-id"]');
    btnChangeId?.dispatchEvent(new Event('click'));

    expect(component.onClientUpdated.emit).toHaveBeenCalledWith({
      id: 3,
      name: 'Juan'
    });
  })

  test('should emit onChangeClient with specific ID if have client', () => {
    jest.spyOn(component.onClientUpdated, 'emit');
    component.onChange(10);
    expect(component.onClientUpdated.emit).not.toHaveBeenCalled();
    component.client = { id: 1, name: 'Juan' };
    fixture.detectChanges();
    component.onChange(10);
    expect(component.onClientUpdated.emit).toHaveBeenCalledWith({
      id: 10,
      name: 'Juan'
    });
  })
});
