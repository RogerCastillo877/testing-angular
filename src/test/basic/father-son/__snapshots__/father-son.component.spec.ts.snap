// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FatherSonComponent should make match with snapshot 1`] = `
<div
  id="root1"
>
  <h4>
    Son
  </h4>
  <code>
    
  </code>
  <br />
  <br />
  
  
</div>
`;
